### PR TITLE
Delete time series data when adding exclusion periods

### PR DESCRIPTION
**Pull request by tleraj**
The purpose of this PR is to close https://github.com/aqualinkorg/aqualink-app/issues/628.
With this implementation, we delete the corresponding time series data whenever a user adds an exclusion period. Also, we take the existing exclusion periods into consideration whenever we run the spotter backfill script, so that it does not backfill data for these periods.

### Designs
| dialog | dates completed | `ADD PERIOD` button clicked |
|---|---|---|
| Screenshot_20220204_165336 | Screenshot_20220204_165406 | Screenshot_20220204_165419 |

### CODE REVIEW

**Comment by ericboucher** [comment]
Let's add a warning popup for confirmation - to make sure users are aware that data will actually be deleted.

File: packages/api/src/sites/sites.service.ts [hunk 1]
```diff
@@ -399,11 +405,44 @@ export class SitesService {
...
+        `Exclusion period [${startDate}, ${endDate}] already exists`,
+      );
+    }
+
+    // We assume that a specific spotter corresponds to only one source.
+    const source = await this.sourceRepository.findOne({
+      where: { sensorId: site.sensorId, type: SourceType.SPOTTER },
+    });
+
+    if (!source) {
```
Comment (ericboucher) [inline packages/api/src/sites/sites.service.ts:423]: If no source is found, it could mean that no data has been uploaded yet but just the spotter just got added. So let's just return early instead, and send back a warning, not an error?

File: packages/api/src/utils/spotter-time-series.ts [hunk 2]
```diff
@@ -77,6 +79,49 @@ const getSpotterSources = (
...
+    exclusionDatesRepository.find({ where: { sensorId: source.sensorId } }),
+  );
+
+/**
+ * Filters out spotter data whose timestamp falls into any exclusion date interval.
+ * @param data The spotter data to filter
+ * @param exclusionDates An array of exclusion dates
+ * @returns The filtered spotter data
+ */
+const excludeSpotterData = (
```
Comment (ericboucher) [inline packages/api/src/utils/spotter-time-series.ts:102]: I have a feeling this function already exists somewhere / or similar. It's great to create util functions, make sure to leverage them in old code as well.

File: packages/api/src/sites/sites.service.ts (hunk 1 above)
Comment (tleraj) [reply packages/api/src/sites/sites.service.ts]: On second thoughts, we could delete the data only if the corresponding source exists. Otherwise, just add the exclusion period on the `exclusion_dates` table, and when the spotter data will be backfilled, this period will be taken into account.
I adopted this logic in https://github.com/aqualinkorg/aqualink-app/pull/668/commits/d04a1466377668b996688b5c8c8a55b002a5f26b.

File: packages/api/src/utils/spotter-time-series.ts (hunk 2 above)
Comment (tleraj) [reply packages/api/src/utils/spotter-time-series.ts]: Leveraged a pre-existing util in https://github.com/aqualinkorg/aqualink-app/pull/668/commits/c09780cd71b53ab21a48e269956a22bcaf19a871.

File: packages/api/src/sites/dto/spotter-data.dto.ts [hunk 3]
```diff
@@ -1,6 +1,6 @@
 import { TimeSeriesValueDto } from '../../time-series/dto/time-series-value.dto';
 
 export class SpotterDataDto {
-  topTemperature: TimeSeriesValueDto[];
-  bottomTemperature: TimeSeriesValueDto[];
+  topTemperature?: TimeSeriesValueDto[];
```
Comment (ericboucher) [inline packages/api/src/sites/dto/spotter-data.dto.ts:4]: is this change needed? Why not return empty lists when needed instead of making it optional?

File: packages/website/src/store/Sites/selectedSiteSlice.ts [hunk 4]
```diff
@@ -213,6 +215,29 @@ const selectedSiteSlice = createSlice({
       }
       return state;
     },
+    excludeSourceData: (
```
Comment (ericboucher) [inline packages/website/src/store/Sites/selectedSiteSlice.ts:218]: Why do we need to do this in the frontend?

**Review by ericboucher — COMMENTED** [review]
Nice work @tleraj, a few open questions but we're getting there

File: packages/website/src/store/Sites/selectedSiteSlice.ts (hunk 4 above)
Comment (tleraj) [reply packages/website/src/store/Sites/selectedSiteSlice.ts:218]: When we make a request to delete time series data, we do not want to immediately to make a new request to fetch (a possibly huge amount) of the updated data, in order to view them on the frontend, so we handle the already stored data on redux store.

File: packages/api/src/sites/dto/spotter-data.dto.ts (hunk 3 above)
Comment (ericboucher) [reply packages/api/src/sites/dto/spotter-data.dto.ts:4]: For spotters we should always get the same format, so I think we should keep it as it was with non optional arrays.

File: packages/website/src/store/Sites/selectedSiteSlice.ts (hunk 4 above)
Comment (ericboucher) [reply packages/website/src/store/Sites/selectedSiteSlice.ts:218]: Let's dispatch a new requests, excluding data should be pretty rare and happen at the beginning of a spotter's life
